rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,578 @@
+/// !REVIEW: I've tried to closely follow SymCrypts API calls and semantics making some simplifications, but I think we have a chance to simplify further if we choose to.
+/// My questions / Suggestions:
+/// 1. We've introduced the idea of the RsaKeyPair / RsaPublicKey state, which can hold a lot of data, and makes a lot of the calls to the underlying SymCrypt code redundant. most of the accessors no longer
+/// needed to call SymCrypt since the same data is stored on the state or is assumed
+/// 2. to keep this thread safe, I've combined the alloc / generate and alloc / set pattern to just one function
+/// 3. This code is pretty raw, and would not be very usable, the idea would be to write a crate on top of this one, called something like symcrypt_rsa_parse, which will be a friendlier
+/// interface on top of the RSA code we have right now it would have convenience functions like rsa_from_pkcs8, rsa_from_der, rsa_from_asn1 etc. ( this could be part of the intern project as well.)
+/// 4. with how many steps / options there is for generating a key, we can maybe do a builder pattern, this would be similar to how it is on NCrypt / BCrypt where they have
+/// RsaKey functions and the RsaKey.finalize() at the end.
+/// ex:
+///         builder = RsaKeyBuilder::new()
+///             .key_mode(RsaKeyMode::KeyPair)
+///             .key_usage(RsaKeyUsage::Sign)
+///             .generate_key()
+/// My thought is to leave it how it is right now as there are not that many parameters that we need to specify, but thought I would point out the pattern.
+///
+/// 5.
+// !REVIEW: instead of using in/out buffers for the get functions, we can use Vec<> which will be slower but more robust, and will not require the user to know the len of the fields,
+//  Or we could instead return a struct with the info
+//       pub struct KeyReturnValue {
+//          key_mode : RsaKeyMode,
+//          modulus: u64,
+//          p: &[u8],
+//          q: &[u8],
+//              ...
+//        }
+
+/// 6. A lot of the functions take in a NumberFormat ( MSB, LSB.) Should we include this when we create a new Public / Key pair state?  Also talked with Phil, ASN is MSB, are there any
+/// use cases for LSB? Should we remove the option from the user and force MSB?
+///
+// Rsa functions. For further documentation please refer to symcrypt.h
+// TODO rest of the documentation
+use crate::errors::SymCryptError;
+use crate::NumberFormat;
+use std::ptr;
+
+/// 2^16 + 1
+const DEFAULT_PUBLIC_EXPONENT: u64 = 0x10001;
+
+/// Rsa Public and Private Key State.
+///
+/// [`RsaKeyPair`] stores a private and public Rsa key.
+pub struct RsaKeyPair {
+    inner: symcrypt_sys::PSYMCRYPT_RSAKEY,
+    key_usage: RsaKeyUsage,
+    number_format: NumberFormat,
```

Comment:
>   number_format: NumberFormat, [](http://example.com/codeflow?start=1&length=31)

Why is this number_format persisted? Its a parameter to symcrypt import or export function.

Reply:
From SymCrypt comments:

> SymCryptRsakeySetValue and SymCryptRsakeyGetValue
> //  Remarks:
>//  - Modulus and all primes are stored in the same format specified by numFormat.

Even though `generate` does not require the number format, the user must  use MSB / LSB for the `pub_exp` when they generate the key 

`pub fn generate_new(
        n_bits_mod: u32,
        pub_exp: Option<&[u8]>,
        key_usage: RsaKeyUsage,
        num_format: NumberFormat,
    ) -> Result<Self, SymCryptError> {
`

Which means that if they supply `pub_exp` as LSB, then later try to export with `MSB` it will be wrong. IMO, having the Number Format at the beginning of the key creation makes the most sense. 

Follow up question that I have is, do we need to include `NumberFormat` at all? It looks like Ncrypt always uses MSB, Is there a scenario where LSB is needed?

Reply:
Internal to symcrypt, its MSB. However, symcrypt supports importing or exporting using LSB.

Reply:
Resolved offline